server/src/main/java/org/opensearch/cluster/routing/WeightedRoundRobin.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class WeightedRoundRobin<T> implements Iterator<T>, Iterable<T> {
+
+    private List<WeightedRoundRobin.Entity<T>> entities;
+    private int turn;
+    private int lastSelectedEntity;
+    private double currentWeight = 0;
+
+    public WeightedRoundRobin(List<WeightedRoundRobin.Entity<T>> entities) {
+        this.entities = entities;
+        this.turn = 0;
+        this.lastSelectedEntity = -1;
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#hasNext()
+     */
+    @Override
+    public boolean hasNext() {
+        return entities.size() > 0;
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#next()
+     */
+    @Override
+    public T next() {
+        Entity<T> entity = entities.get(turn++);
+        return entity.getTarget();
+    }
+
+    /* (non-Javadoc)
+     * @see java.lang.Iterable#iterator()
+     */
+    @Override
+    public Iterator<T> iterator() {
+        return this;
+    }
+
+    /**
+     * *
+     * @return list of entities that is ordered using weighted round-robin scheduling
+     * http://kb.linuxvirtualserver.org/wiki/Weighted_Round-Robin_Scheduling
+     */
+    public List<WeightedRoundRobin.Entity<T>> orderEntities() {
+        int size = entities.size();
+        List<WeightedRoundRobin.Entity<T>> orderedWeight = new ArrayList<>();
+        if (size <= 0) {
+            return null;
+        }
+        if (size == 1) {
+            return entities;
+        }
+        // Find maximum weight and greatest common divisor of weight across all entities
+        double maxWeight = 0;
+        double sumWeight = 0;
+        Double gcd = null;
+        for (WeightedRoundRobin.Entity<T> entity : entities) {
+            maxWeight = Math.max(maxWeight, entity.getWeight());
+            gcd = (gcd == null) ? entity.getWeight() : gcd(gcd, entity.getWeight());
+            sumWeight += entity.getWeight();
+        }
+        int count = 0;
+        while (count < sumWeight) {
+            lastSelectedEntity = (lastSelectedEntity + 1) % size;
+            if (lastSelectedEntity == 0) {
+                currentWeight = currentWeight - gcd;
+                if (currentWeight <= 0) {
+                    currentWeight = maxWeight;
+                    if (currentWeight == 0) {
```

Comment:
Didn't get this. Tried to follow the standard Weighted round robin implementation - http://kb.linuxvirtualserver.org/wiki/Weighted_Round-Robin_Scheduling

Reply:
Will it work if weights are some non 0 &1 numbers like 0.15, 0.25, 0.60?

Reply:
This algo  works for integer weights

Reply:
Ok, Can we call it out explicitly. Even for integers if the values are 3, 6, 9  then also it would just return value 9. Please check once and make an explicit comment on which all scenarios it will work. Use case could be that we may want to give different weights in the future based on node type in case of heterogeneous clusters.